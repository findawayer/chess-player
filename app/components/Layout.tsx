import { Box, CssBaseline } from '@material-ui/core';
import { ThemeProvider } from '@material-ui/core/styles';
import Head from 'next/head';
import type { FunctionComponent, ReactNode } from 'react';

import { ColorModeApiProvider, useColorMode } from '~app/hooks';
import { useMuiTheme } from '~app/themes';
import Header from './Header';

interface LayoutProps {
  children: ReactNode;
}

/** Outmost presentional layer. Should be wrapping all pages. */
const Layout: FunctionComponent<LayoutProps> = ({ children }) => {
  // Local state: Restore color mode from client cache.
  const colorModeApi = useColorMode();
  /** Material UI theme */
  const theme = useMuiTheme(colorModeApi.colorMode);

  // CssBaseline: global CSS created with Material UI.
  return (
    <ThemeProvider theme={theme}>
      <ColorModeApiProvider value={colorModeApi}>
        <Head>
          <meta name="theme-color" content={theme.palette.background.default} />
        </Head>
        <CssBaseline />
        <Box
          display="flex"
          flexDirection="column"
          role="presentation"
          style={{ height: '100%' }}
        >
          <Header />
          <Box
            component="main"
            flexGrow={1}
            display="flex"
            alignItems="center"
            justifyContent="center"
          >
            {children}
          </Box>
        </Box>
      </ColorModeApiProvider>
    </ThemeProvider>
  );
};

export default Layout;
