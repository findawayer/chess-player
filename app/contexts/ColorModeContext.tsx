import { ColorMode } from '@prisma/client';
import { createContext } from 'react';
import { DEFAULT_COLOR_MODE } from '~app/hooks';

type ColorModeContextType = {
  colorMode: ColorMode;
  toggleColorMode(): void;
};

export const ColorModeContext = createContext<ColorModeContextType>({
  colorMode: DEFAULT_COLOR_MODE,
  toggleColorMode: () => null,
});

export const ColorModeProvider = ColorModeContext.Provider;
