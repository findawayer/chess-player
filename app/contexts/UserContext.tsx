import { useQuery } from '@apollo/client';
import React, { createContext } from 'react';

import { CurrentUser, CURRENT_USER_QUERY } from '~app/graphql';

/** React context holding authenticated user payload. */
export const UserContext = createContext<CurrentUser | null>(null);

/** React provider for the user context above. */
export const UserProvider: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => {
  const { data } = useQuery<{ me: CurrentUser | null }>(CURRENT_USER_QUERY);

  return (
    <UserContext.Provider value={data ? data.me : null}>
      {children}
    </UserContext.Provider>
  );
};
