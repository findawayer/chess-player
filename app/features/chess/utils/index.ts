export * from './classNames';
export * from './moves';
export * from './pieces';
export * from './players';
export * from './squares';
