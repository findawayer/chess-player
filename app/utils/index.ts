export * from './color-mode';
export * from './roles';
export * from './testing';
export * from './texts';
