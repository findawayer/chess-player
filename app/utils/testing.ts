import type { User } from '@prisma/client';

const fakeTime = new Date();

export const fakeUser = (overrides?: Partial<User>): User => ({
  id: 1,
  email: 'some@email.com',
  password: '1234',
  name: 'someone',
  role: 'ADMIN',
  createdAt: fakeTime,
  updatedAt: fakeTime,
  verified: false,
  resetToken: null,
  resetTokenExpiry: null,
  colorMode: null,
  chessAutoQueen: null,
  chessBoardColor: null,
  chessShowLegal: null,
  chessShowRecent: null,
  ...overrides,
});
