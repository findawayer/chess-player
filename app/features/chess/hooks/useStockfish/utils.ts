import { Move } from 'chess.js';

// Get the engine's thinking depth allowance.
// (Beyond level 15, let the engine decide with `null` value.)
export const calculateDepth = (skillLevel: number): number | null =>
  skillLevel < 15 ? Math.ceil((skillLevel + 1) / 5) : null;

// Get the probability of playing mistakes.
// NOTE: Stockfish level 20 does not make errors (intentially),
// so these numbers have no effect on level 20.
// Level 0 starts at 1
export const calculateErrorProbability = (skillLevel: number): number =>
  Math.round(skillLevel * 6.35 + 1);

// Level 0 starts at 10
export const calculateMaxError = (skillLevel: number): number =>
  Math.round(skillLevel * -0.5 + 10);

// Format time string that conforms to Stockfish API
export const formatTimeString = ({
  depth,
  duration,
  increment,
}: {
  depth: number | null;
  duration: number;
  increment: number;
}): string => {
  let output = depth ? `depth ${depth}` : '';
  if (duration) {
    output += ` wtime ${duration} winc ${increment} btime ${duration} binc ${increment}`;
  }
  return output;
};

export const formatMoveString = (history: Move[]): string =>
  history.reduce(
    (output, { from, to, promotion }) =>
      `${output} ${from}${to}${promotion ?? ''}`,
    '',
  );
