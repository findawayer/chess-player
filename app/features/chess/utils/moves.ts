import { ChessInstance, Move } from 'chess.js';
import uniqBy from 'lodash/uniqBy';

import { NUMBER_OF_RANKS } from '~app/features/chess/constants';
import {
  ChessLegalMove,
  ChessMoveLog,
  ChessPiece,
  ChessSquare,
  ChessSquareName,
} from '~app/features/chess/types';
import { objectifySquare, stringifySquare } from './squares';

/** Find out whether the chess move passed is a castling. */
export function isCastling(move: Move): boolean {
  return /[kq]/.test(move.flags);
}

/** Find out whether the chess move passed is a promotion. */
export function isPromotion(move: Move): boolean {
  return move.flags.indexOf('p') !== -1;
}

/** Find out whether the chess move passed is an `en passant` capture. */
export function isEnPassant(move: Move): boolean {
  return move.flags.indexOf('e') !== -1;
}

export function getCapturedPiecePosition(move: Move): ChessSquareName | null {
  const match = move.san.match(/x([a-h][0-8])$/);
  return match ? (match[1] as ChessSquareName) : null;
}

export function isPromotingPawn(
  piece: ChessPiece,
  targetSquare: ChessSquare,
): boolean {
  const isPawn = piece.variant === 'p';
  const isWhitePiece = piece.color === 'w';
  const { y } = objectifySquare(targetSquare);
  const isAtTheLastRank = isWhitePiece ? y === NUMBER_OF_RANKS - 1 : y === 0;
  return isPawn && isAtTheLastRank;
}

/** Get list of possible squares and their move type from the square given. */
export function findLegalMoves(
  validator: ChessInstance,
  sourceSqaure: ChessSquare,
): ChessLegalMove[] {
  // Get the legal moves from the validator, and extract payload.
  const legalMoves = validator
    .moves({ square: stringifySquare(sourceSqaure), verbose: true })
    .map(({ to, flags }) => ({ square: to, flags }));
  // Filter duplicate promotion variants.
  return uniqBy(legalMoves, 'square');
}

/** Get the latest move's source and target squares from the move history. */
export function getRecentMovePath(
  moveHistory: ChessMoveLog[],
): ChessSquareName[] {
  if (!moveHistory.length) {
    return [];
  }
  const lastMove = moveHistory[moveHistory.length - 1];
  return [lastMove.from, lastMove.to];
}
