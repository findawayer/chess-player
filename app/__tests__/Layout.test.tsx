import { MockedProvider } from '@apollo/client/testing';
import { render, waitFor } from '@testing-library/react';

import Layout from '~app/components/Layout';
import { UserProvider } from '~app/contexts';
import { CURRENT_USER_QUERY } from '~app/graphql';
import { fakeUser } from '~app/utils';

const anonymousMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { me: null } },
  },
];

const authUserMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { me: fakeUser({ colorMode: 'DARK' }) } },
  },
];

// Mock `window.matchMedia`
// https://jestjs.io/docs/en/manual-mocks#mocking-methods-which-are-not-implemented-in-jsdom
Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: jest.fn().mockImplementation(query => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

describe('<Layout />', () => {
  it('renders properly for anonymous users.', async () => {
    const { container } = render(
      <MockedProvider mocks={anonymousMocks}>
        <UserProvider>
          <Layout>...</Layout>
        </UserProvider>
      </MockedProvider>,
    );
    waitFor(() => expect(container).toMatchSnapshot());
  });

  it('renders properly for anonymous users.', async () => {
    const { container } = render(
      <MockedProvider mocks={authUserMocks}>
        <UserProvider>
          <Layout>...</Layout>
        </UserProvider>
      </MockedProvider>,
    );
    waitFor(() => expect(container).toMatchSnapshot());
  });
});
