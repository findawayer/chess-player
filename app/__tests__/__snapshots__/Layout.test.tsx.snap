// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Layout /> renders properly for anonymous users. 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-1"
    role="presentation"
    style="height: 100%;"
  >
    <header
      class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation2"
    >
      <div
        class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
      >
        <div
          class="MuiTypography-root makeStyles-title-2 MuiTypography-h1 MuiTypography-noWrap"
        >
          <a
            class="makeStyles-titleLink-3"
            href="/"
            rel="home"
          >
            chess-player
          </a>
        </div>
        <div
          class="MuiBox-root MuiBox-root-7 makeStyles-colorModeToggler-4"
        >
          <label
            for="colorModeToggler"
          >
            <span>
              Light
            </span>
            <span
              class="MuiSwitch-root"
            >
              <span
                aria-disabled="false"
                class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-8 MuiSwitch-switchBase"
                title="Toggle theme"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <input
                    class="PrivateSwitchBase-input-11 MuiSwitch-input"
                    id="colorModeTogler"
                    type="checkbox"
                    value=""
                  />
                  <span
                    class="MuiSwitch-thumb"
                  />
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </span>
              <span
                class="MuiSwitch-track"
              />
            </span>
            <span>
              Dark
            </span>
          </label>
        </div>
        <div
          class="makeStyles-menu-5"
        >
          <a
            aria-disabled="false"
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-colorInherit"
            href="https://github.com/findawayer/chess-player"
            rel="external noopener noreferrer"
            tabindex="0"
            target="_blank"
          >
            <span
              class="MuiButton-label"
            >
              GitHub
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
          <a
            aria-disabled="false"
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-colorInherit"
            href="/get-started"
            tabindex="0"
          >
            <span
              class="MuiButton-label"
            >
              Join
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
          <a
            aria-disabled="false"
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-colorInherit"
            href="/signin"
            tabindex="0"
          >
            <span
              class="MuiButton-label"
            >
              Login
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </a>
        </div>
      </div>
    </header>
    <main
      class="MuiBox-root MuiBox-root-6"
    >
      ...
    </main>
  </div>
</div>
`;

exports[`<Layout /> renders properly for anonymous users. 2`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-12"
    role="presentation"
    style="height: 100%;"
  >
    <header
      class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation2"
    >
      <div
        class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
      >
        <div
          class="MuiTypography-root makeStyles-title-13 MuiTypography-h1 MuiTypography-noWrap"
        >
          <a
            class="makeStyles-titleLink-14"
            href="/"
            rel="home"
          >
            chess-player
          </a>
        </div>
        <div
          class="makeStyles-menu-16"
        >
          <a
            aria-disabled="false"
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-colorInherit"
            href="https://github.com/findawayer/chess-player"
            rel="external noopener noreferrer"
            tabindex="0"
            target="_blank"
          >
            <span
              class="MuiButton-label"
            >
              GitHub
            </span>
          </a>
          <a
            aria-disabled="false"
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-colorInherit"
            href="/get-started"
            tabindex="0"
          >
            <span
              class="MuiButton-label"
            >
              Join
            </span>
          </a>
          <a
            aria-disabled="false"
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-colorInherit"
            href="/signin"
            tabindex="0"
          >
            <span
              class="MuiButton-label"
            >
              Login
            </span>
          </a>
        </div>
      </div>
    </header>
    <main
      class="MuiBox-root MuiBox-root-17"
    >
      ...
    </main>
  </div>
</div>
`;
